### PR TITLE
Make variables configurable

### LINKED ISSUE #476: Add channel selection capability to data loaders
### Problem / Goal

Every input dataset group currently contributes all of its variables as model input channels. In `CommonDataModule.input_spaces` we build one `DataSpace` per group via `SingleDataset.space`, which uses `channels = dataslices[0].shape[1]`

Instead of this, we should allow variable subsetting, as is done for the target group in `CommonDataModule._target_variables`.

## Suggested mechanism

- Extend `SingleDataset.subset(variables=...)` usage to input groups, not just the target.
- Add config, e.g. inputs.<group>.variables: [...], parallel to the existing predict.target.variables pattern
- In `CommonDataModule` call `ds.subset(variables=...)` on each input `SingleDataset` before it's wrapped by `CombinedDataset`

## Testing

- Unit test: SingleDataset.subset applied to an input group changes DataSpace.channels and the encoder built for that space.
- Compose check: a config with `inputs.<group>.variables` set composes and the model builds an encoder with the reduced channel count
- Confirm no interaction with `predict.target.variables`

### What does good look like?

- [ ] Able to select individual variables from input datasets

### Related context

_No response_

### PR DESCRIPTION
**Pull request by jemrobinson**
## Summary

Make both input and target variables configurable in one place.

### Config changes

- `variables`: new config group describing input and target variables (replacing `predict.target.variables` on the target side)
- `window`: new config group containing just `batch_size`, `n_forecast_steps` and `n_history_steps` that provides a single place to define the `NT` dimensions of our tensors
- `predict`: config group is obsoleted

### Data loading
- `CommonDataModule` now only loads the specific variables requested under `variables.input`/`variables.target` per dataset group, rather than every variable in the underlying zarr.
- Validation of dataset groups and variable names moved to first `cached_property` access 
- `CombinedDataset.variable_list()` added for logging which variables a constructed dataset actually contains.
- Simplified `_dataloader` methods in `CommonDataModule`

### Plotting fix
- `BaseModel` now accepts and stores `channel_names`, populated from `data_module.target_variables`, which fixes W&B plot file names
- `EncoderStage` drops its redundant `channel_names` handling

Closes #476. Resolves #497.

### CODE REVIEW

File: icenet_mp/data/common_data_module.py [hunk 1]
```diff
@@ -94,7 +80,19 @@ def __init__(self, config: DictConfig) -> None:
...
     @cached_property
     def datasets(self) -> dict[str, SingleDataset]:
-        """Return a dictionary of dataset group names to SingleDataset objects."""
+        """Return a filtered dictionary of dataset group names to SingleDataset objects.
+
+        Only include requested variables for each dataset group. If no variables are
+        requested for a dataset group, ignore it.
+        """
+        return {
+            name: self.datasets_unfiltered[name].subset(variables=variables)
```
Comment (AoifeHughes) [inline icenet_mp/data/common_data_module.py:89]: I think the co-pilot comment below is meaning to tag this one? It's ordering is a bug? variables are filtered which could mess up things based on how anemoi keeps ordered sets? So would something like `subset` fix?

File: icenet_mp/data/common_data_module.py [hunk 2]
```diff
@@ -160,111 +158,153 @@ def output_space(self) -> DataSpace:
...
     def target_variable_indices(self) -> list[int]:
-        """Return the indices of the variables to predict."""
-        return [
-            self.variable_names[self.target_group_name].index(variable)
-            for variable in self.target_variables
-        ]
+        """Return the indices of the target variables within their dataset group."""
+        try:
+            return [
+                self.variable_names[self.target_group_name].index(variable)
```
Comment (AoifeHughes) [inline icenet_mp/data/common_data_module.py:208]: `target_variable_indices` indexes with `.index()` against the config-ordered `variable_names` built here, so the indices are config order while the data they index into (per the row above) is zarr order. shipped configs get away with it, single target, and the mock zarr order happens to agree, so `test_input_selection_changes_data_space_channels` only passes because the mock's variable order (`ice_conc, ice_thickness, temperature`) happens to match the requested order for that test. need a test where requested order != zarr order

File: icenet_mp/model_service.py [hunk 3]
```diff
@@ -72,6 +72,7 @@ def from_config(cls, config: DictConfig) -> "ModelService":
         log.info("Building a new '%s' model...", builder.config["model"]["_target_"])
         builder.model_ = hydra.utils.instantiate(
             config["model"],
+            channel_names=builder.data_module.target_variables,
```
Comment (AoifeHughes) [inline icenet_mp/model_service.py:75]: new line, `channel_names=builder.data_module.target_variables` for the main model, config order. once channel order and data order disagree, W&B labels go with them. fixing the `subset` ordering fixes this for free

File: icenet_mp/model_service.py [hunk 4]
```diff
@@ -104,15 +105,15 @@ def from_checkpoint(
...
         try:
             # Load the model configuration from the checkpoint directory
             ckpt_config = DictConfig(OmegaConf.load(config_path))
             log.debug("Loaded checkpoint configuration from %s.", config_path)
             combined_cfg = DictConfig(OmegaConf.merge(ckpt_config, config))
-            for key in ("model", "predict", "train"):
+            for key in ("model", "train", "window"):
                 combined_cfg[key] = OmegaConf.merge(
                     combined_cfg.get(key, {}), ckpt_config.get(key, {})
                 )
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/model_service.py:120]: Done in c5c0a0c6

File: icenet_mp/data/common_data_module.py [hunk 5]
```diff
@@ -94,7 +80,19 @@ def __init__(self, config: DictConfig) -> None:
...
-        """Return a dictionary of dataset group names to SingleDataset objects."""
+        """Return a filtered dictionary of dataset group names to SingleDataset objects.
+
+        Only include requested variables for each dataset group. If no variables are
+        requested for a dataset group, ignore it.
+        """
+        return {
+            name: self.datasets_unfiltered[name].subset(variables=variables)
+            for name, variables in self.variable_names.items()
+        }
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/data/common_data_module.py:115]: Fixed in d8f975b0

File: tests/config/test_hydra.py [hunk 6]
```diff
@@ -14,7 +14,7 @@ def test_sample_config_has_expected_top_level_keys(
         self, compose_config: Callable[..., DictConfig]
     ) -> None:
         cfg = compose_config()
-        for key in ("data", "model", "train", "loss", "predict", "evaluate"):
+        for key in ("data", "evaluate", "loss", "model", "train", "window"):
             assert key in cfg, f"Key '{key}' missing from composed config"
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply tests/config/test_hydra.py]: Fixed in 8a40ab10

File: tests/data/test_common_data_module.py [hunk 7]
```diff
@@ -38,71 +84,280 @@ def test_mixed_none_and_string_in_same_period(
...
+                "ds1": {"name": mock_dataset.stem, "group_as": "group1"},
+                "ds2": {"name": mock_dataset.stem, "group_as": "group2"},
+            },
+            input_variables={"group2": ["ice_conc"]},
+            target_variables={"group1": ["ice_conc"]},
         )
+        dm = CommonDataModule(cfg)
+
+        with pytest.raises(StopIteration):
+            _ = dm.target_variables
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply tests/data/test_common_data_module.py]: Fixed in d8f975b

File: tests/conftest.py [hunk 8]
```diff
@@ -186,18 +188,22 @@ def cfg_model_service() -> DictConfig:
...
                 "callbacks": {},
                 "optimizer": {},
                 "scheduler": {},
                 "lr_scheduler": {},
                 "trainer": {},
             },
+            "variables": {
+                "input": {"mock-dataset-group-1": ["mock_var"]},
+                "output": {"mock-dataset-group-1": ["mock_var"]},
+            },
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply tests/conftest.py:255]: Done in 62856be

File: docs/src/user-guide/configuration.md [hunk 9]
```diff
@@ -87,6 +87,28 @@ And run with:
...
+  sic-osisaf:
+    - ice_conc
+target:
+  sic-osisaf:
+    - ice_conc
+```
+
+!!! note
+    We current enforce that the target variables must also be input variables.
+    This requirement may be relaxed in future.
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply docs/src/user-guide/configuration.md:110]: Done in 96fca716

File: icenet_mp/data/common_data_module.py (hunk 2 above)
Comment (jemrobinson) [reply icenet_mp/data/common_data_module.py]: Great catch! Fixed in 71d98388

File: icenet_mp/data/common_data_module.py (hunk 1 above)
Comment (jemrobinson) [reply icenet_mp/data/common_data_module.py]: I think this is done in 0a5efad4. Do you agree?

File: icenet_mp/model_service.py (hunk 3 above)
Comment (jemrobinson) [reply icenet_mp/model_service.py:75]: Should also be fixed by 0a5efad4

File: icenet_mp/config/variables/synthetic.yaml [hunk 10]
```diff
@@ -0,0 +1,38 @@
+input:
+  era5:
+    - 10u
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/config/variables/synthetic.yaml]: Done in 74ccb0d3

File: icenet_mp/config/base.yaml [hunk 11]
```diff
@@ -4,9 +4,10 @@ defaults:
   - evaluate: default
   - loss: amse
   - model: quick_test
-  - predict: sic-osisaf-2d
   - random: default
   - reporting: default
   - train: default
   - platform: default
+  - variables: sic-osisaf
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/config/base.yaml:11]: Done in 74ccb0d3

File: icenet_mp/data/common_data_module.py [hunk 12]
```diff
@@ -169,101 +250,145 @@ def output_space(self) -> DataSpace:
...
+                ds.variable_names
+                for ds in self.datasets.values()
+                if ds.name == self.target_group_name and ds.variable_names
+            )
+        except StopIteration as exc:
+            msg = f"Dataset group {self.target_group_name} has no available variables."
+            raise ValueError(msg) from exc
+        # Verify that the requested variable names exist in the dataset group
+        requested_variables = self._requested_target_variables[self.target_group_name]
+        for requested_variable in requested_variables:
```
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/data/common_data_module.py]: Done in 2435304e

File: icenet_mp/model_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jemrobinson) [reply icenet_mp/model_service.py:75]: Ignoring this one as it is likely to conflict with #544 .
